### PR TITLE
New gateway

### PR DESCRIPTION
**Pull request by gakonst**
- Bindings for the upgraded gateway contract.
- Adds optional version flag during creation of the clients which is used to use the new signing method when withdrawing.

### CODE REVIEW

File: src/dpos-user.ts [hunk 1]
```diff
@@ -379,20 +397,62 @@ export class DPOSUser {
...
+      sig = sig.length === 134 ? '0x' + sig.slice(4) : sig
+      let sign = ethers.utils.splitSignature(sig)
+
+      let signer = ethers.utils.recoverAddress(
+        ethers.utils.arrayify(ethers.utils.hashMessage(ethers.utils.arrayify(hash))),
+        sign
+      )
+      log('Receipt was signed by:', signer)
+
+      let valIndexes = [0]
```
Comment (gakonst) [inline src/dpos-user.ts]: Note that in a future PR this was edited to be calculated based on the current validators list from the ValidatorManagerContract. The current task does not require it yet so I prefer to get this in as is, and complete it later. For reference, this is what's going to happen: https://github.com/loomnetwork/loom-js/pull/216/commits/ee0b806f29b44002516a0d58df9c9c07a11cd2b5

File: src/dpos-user.ts [hunk 2]
```diff
@@ -20,23 +20,27 @@ import {
 const log = debug('dpos-user')
 
 const coinMultiplier = new BN(10).pow(new BN(18))
-const ERC20GatewayABI = require('./mainnet-contracts/ERC20Gateway.json')
+const V2_GATEWAYS = ['oracle-dev', 'asia1']
```
Comment (enlight) [inline src/dpos-user.ts]: Err, no, we're not going to hardcode chain IDs in loom-js, can hardcode it in whatever uses loom-js.

File: src/dpos-user.ts [hunk 3]
```diff
@@ -20,23 +20,27 @@ import {
...
   private _address: Address
   private _ethAddress: string
-  private _ethereumGateway: ERC20Gateway
+  private _ethereumGateway: ERC20Gateway_v2
   private _ethereumLoom: ERC20
   private _dappchainGateway: Contracts.LoomCoinTransferGateway
   private _dappchainLoom: Contracts.Coin
   private _dappchainDPOS: Contracts.DPOS2
   private _dappchainMapper: Contracts.AddressMapper
+  private _version: number
```
Comment (enlight) [inline src/dpos-user.ts]: Use an `enum` instead of a number.

File: src/dpos-user.ts [hunk 4]
```diff
@@ -394,20 +420,62 @@ export class DPOSUser {
...
 
+    log('Got receipt', pendingReceipt)
     return CryptoUtils.bytesToHexAddr(signature)
   }
 
   private async withdrawCoinFromRinkebyGatewayAsync(
     amount: BN,
     sig: string
   ): Promise<ethers.ContractTransaction> {
+    const hash = await this.createWithdrawalHash(amount)
```
Comment (enlight) [inline src/dpos-user.ts]: Hash is only used when `version == 2` so why is it computed when `version == 1`?

File: src/dpos-user.ts [hunk 5]
```diff
@@ -394,20 +420,62 @@ export class DPOSUser {
...
+    log('Receipt hash:', hash)
+
+    if (this._version === 2) {
+      // Ugly hack to extract the 'mode' bit from the old signature format - if it's still used (68 = 66 + 2, where 2 is the 0x)
+      sig = sig.length === 134 ? '0x' + sig.slice(4) : sig
```
Comment (enlight) [inline src/dpos-user.ts]: Not clear to me where the 134 nibbles come from... `'0x' + 66 * 2`?

File: src/dpos-user.ts [hunk 6]
```diff
@@ -394,20 +420,62 @@ export class DPOSUser {
...
+      let sign = ethers.utils.splitSignature(sig)
```
Comment (enlight) [inline src/dpos-user.ts]: `sign`->`expandedSig`

File: src/dpos-user.ts (hunk 2 above)
Comment (gakonst) [reply src/dpos-user.ts]: That was Matt's request after our discussion to simplify the process by which clients upgrade their logic.

Comment (enlight) [reply src/dpos-user.ts]: Of course it was.

File: src/dpos-user.ts (hunk 5 above)
Comment (gakonst) [reply src/dpos-user.ts]: 66 bytes are represented as a string of length 132
0x is also 2 characters

(signature of 66 bytes) + (0x) = 66 * 2 + 2

File: src/dpos-user.ts (hunk 4 above)
Comment (gakonst) [reply src/dpos-user.ts]: Good point, thanks.

File: src/dpos-user.ts [hunk 7]
```diff
@@ -118,14 +147,17 @@ export class DPOSUser {
...
-    dappchainMapper: Contracts.AddressMapper
+    dappchainMapper: Contracts.AddressMapper,
+    version: GatewayVersion = GatewayVersion.SINGLESIG
   ) {
+    this._version = version
     this._wallet = wallet
     this._address = address
     this._ethAddress = ethAddress
     this._client = client
+    const gatewayABI = version == 2 ? ERC20GatewayABI_v2 : ERC20GatewayABI
```
Comment (enlight) [inline src/dpos-user.ts]: `version === GatewayVersion.MULTISIG`

**Pull request merged**